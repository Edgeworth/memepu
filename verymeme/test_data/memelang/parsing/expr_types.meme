enum Unit {}
fn bubblesort<T>(v: const* vector<*T>) Unit {}
struct vector<T> {
  fn push(v: i32) Unit {}
}
fn main() i32 {
  var v: auto = vector<i8>.new();
  v.push(1);
  bubblesort(&v);
  var d: auto = vector<i16>(v);
  a * *vector<i16>(v);
  return 0;
}
