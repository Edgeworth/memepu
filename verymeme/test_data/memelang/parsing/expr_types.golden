File
 FnDefn
  FnSig
   Typename(bubblesort)
    Typelist(T)
   VarDecl
    VarRef(v)
    Type(vector)
     Qualifier(const: 1 ptr: 1, array: 0)
     Type(T)
      Qualifier(const: 0 ptr: 1, array: 0)
   Type(Unit)
  StmtBlk
 FnDefn
  FnSig
   Typename(main)
   Type(i32)
  StmtBlk
   Var
    VarDecl
     VarRef(v)
     Type(auto)
    Op(FN_CALL)
     FnCallArgs
     Op(MEMBER_ACCESS)
      VarRef(new)
      Type(vector)
       Type(i8)
   Op(FN_CALL)
    FnCallArgs
     IntLit(1)
    Op(MEMBER_ACCESS)
     VarRef(push)
     VarRef(v)
   Op(FN_CALL)
    FnCallArgs
     VarRef(&v)
    VarRef(bubblesort)
   Var
    VarDecl
     VarRef(d)
     Type(auto)
    Op(FN_CALL)
     FnCallArgs
      VarRef(v)
     Type(vector)
      Type(i16)
   Op(MUL)
    Op(UNARY_DEREF)
     Op(FN_CALL)
      FnCallArgs
       VarRef(v)
      Type(vector)
       Type(i16)
    VarRef(a)
   Return
    IntLit(0)
 Enum(0 untyped)
  Typename(Unit)
 Struct
  Typename(vector)
   Typelist(T)
  FnDefn
   FnSig
    Typename(push)
    VarDecl
     VarRef(v)
     Type(i32)
    Type(Unit)
   StmtBlk
